vault/provider_test.go

```diff
@@ -677,19 +754,38 @@ func testTokenName_check(expectedTokenName string) resource.TestCheckFunc {
 			return fmt.Errorf("resource has no primary instance")
 		}
 
-		tokenName, ok := resourceState.Primary.Attributes["data.display_name"]
+		actualValue, ok := resourceState.Primary.Attributes["data."+attrName]
 		if !ok {
-			return fmt.Errorf("cannot access token [%s] for check", "display_name")
+			return fmt.Errorf("cannot access attribute [%s] for check", attrName)
 		}
 
-		if tokenName != expectedTokenName {
-			return fmt.Errorf("token name [%s] expected, but got [%s]", expectedTokenName, tokenName)
+		if actualValue != expectedValue {
+			return fmt.Errorf("%s [%s] expected, but got [%s]", attrName, expectedValue, actualValue)
 		}
 
 		return nil
 	}
 }
 
+func checkUsedToken(expectChildToken bool) resource.TestCheckFunc {
```

Comment:
I think this logic could probably be moved to the test, which would then just call `checkSelfToken()`.

Reply:
I'm not sure what you are asking for here. Can you elaborate, please?

In one interpretation I take your meaning to be that this logic can live in `CheckSelfToken` directly, but because that function is used by other tests that need to check their own specific attributes with their own expectations, this conditional based on whether we are using the child token or not doesn't apply to those tests. In another interpretation I think you may be asking for this to be an anonymous function within either the test case struct, or the `Check` of the test step, but I would argue using an anonymous function in that setting is less clear and readable.

Let me know if I'm missing your meaning.

Reply:
Rather than creating a new function call `checkSelfToken()` directly with the source attribute and the expected value.

Reply:
Could you provide some example code to illustrate how you expect this to look and work please? I must be missing something. Right now the attribute and expected value are based on whether we expect a child token should have been created or not; thus this if-else case. Since we do not know how to call `checkSelfToken` without switching on this condition, this logic needs to occur somewhere, rather than calling "directly" to the `checkSelfToken` function.

Reply:
I was suggesting that we move `checkUsedToken` to be a closure within the related test and renaming it to `checkTokenUsed`:

```
@@ -628,6 +629,19 @@ func TestAccTokenName(t *testing.T) {
 
 func TestAccChildToken(t *testing.T) {
        defer os.Unsetenv("VAULT_SKIP_CHILD_TOKEN")
+
+       checkTokenUsed := func(expectChildToken bool) resource.TestCheckFunc {
+               if expectChildToken {
+                       // If the default child token was created, we expect the token
+                       // used by the provider was named the default "token-terraform"
+                       return checkSelfToken("display_name", "token-terraform")
+               } else {
+                       // If the child token setting was disabled, the used token
+                       // should match the user-provided VAULT_TOKEN
+                       return checkSelfToken("id", os.Getenv("VAULT_TOKEN"))
+               }
+       }
+
        tests := []struct {
                skipChildTokenEnv    string
                useChildTokenEnv     bool
@@ -701,7 +715,7 @@ func TestAccChildToken(t *testing.T) {
                                                }
                                        },
                                        Config: getTerraformConfig(test.useChildTokenSchema, `skip_child_token = `+test.skipChildTokenSchema),
-                                       Check:  checkUsedToken(test.expectChildToken),
+                                       Check:  checkTokenUsed(test.expectChildToken),
                                },
                        },
                })
@@ -762,18 +776,6 @@ func checkSelfToken(attrName string, expectedValue string) resource.TestCheckFun
        }
 }
 
-func checkUsedToken(expectChildToken bool) resource.TestCheckFunc {
-       if expectChildToken {
-               // If the default child token was created, we expect the token
-               // used by the provider was named the default "token-terraform"
-               return checkSelfToken("display_name", "token-terraform")
-       } else {
-               // If the child token setting was disabled, the used token
-               // should match the user-provided VAULT_TOKEN
-               return checkSelfToken("id", os.Getenv("VAULT_TOKEN"))
-       }
-}
-
```

Reply:
I'm happy with that. Done.